File: tests/python/test_vti.py
"""VTK XML ImageData.

The writer requires a lattice, so the fixture set is small on purpose: there is
exactly one shape this format can hold. What is worth testing instead is that the
two implementations agree on that shape's *geometry* -- ``Origin``, ``Spacing``
and ``WholeExtent`` are three attributes that a reader and a writer sharing a
convention can both get wrong together -- and that hand-written files whose
expected geometry is stated independently read correctly.
"""

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus._grid import lattice_from_mesh
from meshioplusplus.vti import _vti


def _lattice(n=3, origin=-0.5, spacing=0.25, point_data=True, cell_data=True):
    m = meshioplusplus.grid([n, n, n], (origin,) * 3, (spacing,) * 3)
    if point_data:
        m.point_data["f"] = np.arange(len(m.points), dtype=np.float64) * 0.5
    if cell_data:
        m.cell_data["tag"] = [np.arange(n**3, dtype=np.int64)]
    return m


@pytest.mark.parametrize(
    "binary, compression", [(False, None), (True, None), (True, "zlib")]
)
def test_roundtrip(binary, compression, tmp_path):
    mesh = _lattice()
    path = tmp_path / "g.vti"
    meshioplusplus.vti.write(path, mesh, binary=binary, compression=compression)
    back = meshioplusplus.vti.read(path)
    assert np.allclose(back.points, mesh.points, atol=1e-12)
    assert len(back.cells) == 1
    assert np.array_equal(back.cells[0].data, mesh.cells[0].data)
    assert np.allclose(back.point_data["f"], mesh.point_data["f"], atol=1e-9)
    assert np.array_equal(back.cell_data["tag"][0], mesh.cell_data["tag"][0])


@pytest.mark.parametrize(
    "binary, compression", [(False, None), (True, None), (True, "zlib")]
)
def test_cross_compat(binary, compression, tmp_path):
    """Each writer's bytes read by the *other* implementation.

    The real cross-check on the base64 block framing: ``_vti._encode_binary`` is
    transcribed from VTU's writer closures, and this is what proves the
    transcription rather than a comment claiming it.
    """
    mesh = _lattice()
    cpp = tmp_path / "cpp.vti"
    meshioplusplus.vti.write(cpp, mesh, binary=binary, compression=compression)
    from_py = _vti.read(cpp)
    assert np.allclose(from_py.points, mesh.points, atol=1e-12)
    assert np.allclose(from_py.point_data["f"], mesh.point_data["f"], atol=1e-9)

    py = tmp_path / "py.vti"
    _vti.write(py, mesh, binary=binary, compression=compression)
    from_cpp = meshioplusplus.vti.read(py)
    assert np.allclose(from_cpp.points, mesh.points, atol=1e-12)
    assert np.allclose(from_cpp.point_data["f"], mesh.point_data["f"], atol=1e-9)
    assert np.array_equal(from_cpp.cell_data["tag"][0], mesh.cell_data["tag"][0])


def test_generic_io(tmp_path):
    """Registration: reachable through ``meshioplusplus.read``/``write``."""
    mesh = _lattice(point_data=False, cell_data=False)
    path = tmp_path / "g.vti"
    meshioplusplus.write(path, mesh)
    back = meshioplusplus.read(path)
    assert np.allclose(back.points, mesh.points)
    assert meshioplusplus.sniff_format(path) == "vti"


def test_reads_a_foreign_file_with_a_shifted_extent(tmp_path):
    """A hand-written file whose expected geometry is stated here, not produced.

    An extent that does not start at zero puts the lo corner at
    ``Origin + start * Spacing``; dropping that offset translates the whole grid
    by one cell, silently and plausibly.
    """
    path = tmp_path / "foreign.vti"
    path.write_text(
        '<?xml version="1.0"?>\n'
        '<VTKFile type="ImageData" version="0.1" byte_order="LittleEndian">\n'
        '<ImageData WholeExtent="2 4 0 1 0 1" Origin="1 2 3" Spacing="0.5 2 4">\n'
        '<Piece Extent="2 4 0 1 0 1">\n'
        '<PointData><DataArray type="Int32" Name="i" format="ascii">\n'
        "0 1 2 3 4 5 6 7 8 9 10 11\n</DataArray></PointData>\n"
        "</Piece></ImageData></VTKFile>\n"
    )
    for reader in (meshioplusplus.vti.read, _vti.read):
        m = reader(path)
        assert len(m.points) == 12
        assert len(m.cells) == 1 and len(m.cells[0].data) == 2
        assert np.allclose(m.points[0], [2.0, 2.0, 3.0])
        assert np.allclose(m.points[1], [2.5, 2.0, 3.0])  # x fastest
        assert np.allclose(m.points[3], [2.0, 4.0, 3.0])  # next y row
        assert np.array_equal(m.point_data["i"], np.arange(12))


def test_refuses_a_mesh_that_is_not_a_lattice(tmp_path):
    path = tmp_path / "no.vti"
    tetra = meshioplusplus.Mesh(
        np.array([[0.0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1]]),
        [("tetra", np.array([[0, 1, 2, 3]]))],
    )
    with pytest.raises(meshioplusplus.WriteError):
        _vti.write(path, tetra)
    # A PARTIAL lattice is the interesting refusal: right cells, right places,
    # and ImageData still cannot express its holes.
    surf = meshioplusplus.extract_surface(meshioplusplus.grid([2, 2, 2]))
    partial = meshioplusplus.voxelize(surf, resolution=(4, 4, 4), fill="surface")
    assert lattice_from_mesh(partial) is None
    with pytest.raises(meshioplusplus.WriteError):
        _vti.write(path, partial)


@pytest.mark.parametrize(
    "text",
    [
        # a piece that is not the whole extent
        '<VTKFile type="ImageData"><ImageData WholeExtent="0 2 0 1 0 1">'
        '<Piece Extent="0 1 0 1 0 1"/></ImageData></VTKFile>',
        # a rotated lattice
        '<VTKFile type="ImageData"><ImageData WholeExtent="0 1 0 1 0 1" '
        'Direction="0 1 0 -1 0 0 0 0 1"><Piece Extent="0 1 0 1 0 1"/>'
        "</ImageData></VTKFile>",
        # the wrong dataset type
        '<VTKFile type="UnstructuredGrid"><UnstructuredGrid/></VTKFile>',
    ],
)
def test_declines_what_it_does_not_implement(text, tmp_path):
    path = tmp_path / "bad.vti"
    path.write_text(text)
    with pytest.raises(meshioplusplus.ReadError):
        _vti.read(path)


def test_the_recovered_header_equals_the_written_one(tmp_path):
    """The identity the format exists for.

    No format persists arbitrary ``field_data``, so a grid written anywhere else
    loses its ``sdf:*`` header. ``.vti`` stores the geometry itself, and
    ``lattice_from_mesh`` reads it back off the points -- so the round trip is
    exact rather than approximate.
    """
    mesh = meshioplusplus.grid([5, 5, 5], (1.25,) * 3, (0.125,) * 3)
    before = lattice_from_mesh(mesh)
    assert before is not None
    path = tmp_path / "g.vti"
    meshioplusplus.vti.write(path, mesh)
    after = lattice_from_mesh(meshioplusplus.vti.read(path))
    assert after is not None
    for a, b in zip(after, before):
        assert np.allclose(a, b, rtol=0, atol=1e-15)


def test_lattice_from_mesh_cpp_matches_python():
    """The recovery agrees with ``detail::lattice_from_mesh``.

    Checked through the .vti writer, which is the C++ side's only consumer: if
    the two disagreed about a mesh's origin or spacing, the file this writes
    would place the grid somewhere the Python twin does not.
    """
    import os
    import tempfile

    mesh = meshioplusplus.grid([3, 4, 5], (-1.5, 0.25, 2.0), (0.5, 0.25, 0.125))
    dims, origin, spacing = lattice_from_mesh(mesh)
    fd, path = tempfile.mkstemp(suffix=".vti")
    os.close(fd)
    try:
        meshioplusplus.vti.write(path, mesh, binary=False)
        text = open(path).read()
    finally:
        os.unlink(path)
    assert f'WholeExtent="0 {dims[0]} 0 {dims[1]} 0 {dims[2]}"' in text

    def num(v):
        return f"{float(v):.17g}"

    assert f'Origin="{num(origin[0])} {num(origin[1])} {num(origin[2])}"' in text
    assert f'Spacing="{num(spacing[0])} {num(spacing[1])} {num(spacing[2])}"' in text

    # And the Python writer writes the identical attribute line.
    fd, path = tempfile.mkstemp(suffix=".vti")
    os.close(fd)
    try:
        _vti.write(path, mesh, binary=False)
        py_text = open(path).read()
    finally:
        os.unlink(path)
    marker = "<ImageData "
    assert (
        text[text.index(marker) : text.index(">", text.index(marker))]
        == py_text[py_text.index(marker) : py_text.index(">", py_text.index(marker))]
    )


def test_lattice_from_mesh_rejects_what_is_not_one():
    tetra = meshioplusplus.Mesh(
        np.array([[0.0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1]]),
        [("tetra", np.array([[0, 1, 2, 3]]))],
    )
    assert lattice_from_mesh(tetra) is None

    # A permuted lattice: identical plane sets, different mesh. A recovery that
    # only looked at the distinct coordinate values would accept this.
    m = meshioplusplus.grid([2, 2, 2])
    permuted = meshioplusplus.Mesh(
        m.points[::-1].copy(), [("hexahedron", m.cells[0].data)]
    )
    assert lattice_from_mesh(permuted) is None

    # A graded grid: uniform plane counts, non-uniform spacing.
    xs = np.array([0.0, 1.0, 3.0])
    pts = np.array(
        [
            [xs[i], float(j), float(k)]
            for k in range(3)
            for j in range(3)
            for i in range(3)
        ]
    )
    graded = meshioplusplus.Mesh(pts, [("hexahedron", m.cells[0].data)])
    assert lattice_from_mesh(graded) is None

File: tests/cpp/test_vti.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// VTK XML ImageData.
//
// A round trip through our own writer is a WEAK oracle here and deliberately not
// the only one: the reader and the writer share a coordinate convention, so a
// consistently wrong `Origin` or a transposed extent survives one unchanged.
// The tests below therefore also (a) read hand-written files whose expected
// geometry is stated independently -- a non-zero Origin, an extent not starting
// at zero, an inverted extent, a foreign Piece -- and (b) assert on the raw
// written bytes.

// System includes
#include <array>
#include <cstddef>
#include <cstdint>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/detail/grid_lattice.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/vti.hpp"
#include "meshioplusplus/operations/voxelize.hpp"

using meshioplusplus::Mesh;
using meshioplusplus::NDArray;
using meshioplusplus::read_vti;
using meshioplusplus::read_vti_metadata;
using meshioplusplus::ReadError;
using meshioplusplus::write_vti;
using meshioplusplus::WriteError;
namespace d = meshioplusplus::detail;

namespace {

std::string vti_write_temp(const std::string& rText) {
    const std::string path = mt::temp_path(".vti");
    std::ofstream os(path, std::ios::binary);
    os << rText;
    os.close();
    return path;
}

std::string vti_slurp(const std::string& rPath) {
    std::ifstream in(rPath, std::ios::binary);
    std::ostringstream ss;
    ss << in.rdbuf();
    return ss.str();
}

Mesh vti_lattice(std::int64_t N, double Origin, double Spacing) {
    return meshioplusplus::grid({{N, N, N}}, {{Origin, Origin, Origin}},
                                {{Spacing, Spacing, Spacing}});
}

}  // namespace

TEST(Vti, RoundTripsALatticeExactly) {
    for (bool binary : {false, true}) {
        Mesh m = vti_lattice(3, -0.5, 0.25);
        NDArray pd = NDArray::Uninit(meshioplusplus::DType::Float64, {m.NumPoints()});
        for (std::size_t p = 0; p < m.NumPoints(); ++p)
            pd.As<double>()[p] = static_cast<double>(p) * 0.5;
        m.AddPointData("f", std::move(pd));
        NDArray cd = NDArray::Uninit(meshioplusplus::DType::Int64, {27u});
        for (std::size_t c = 0; c < 27; ++c)
            cd.As<std::int64_t>()[c] = static_cast<std::int64_t>(c);
        std::vector<NDArray> blocks;
        blocks.push_back(std::move(cd));
        m.AddCellData("tag", std::move(blocks));

        const std::string path = mt::temp_path(".vti");
        write_vti(path, m, binary, /*zlib=*/false);
        const Mesh back = read_vti(path);

        ASSERT_EQ(back.NumPoints(), m.NumPoints());
        ASSERT_EQ(back.NumCellBlocks(), 1u);
        EXPECT_EQ(back.Cells(0).NumCells(), 27u);
        for (std::size_t p = 0; p < m.NumPoints(); ++p)
            for (std::size_t k = 0; k < 3; ++k)
                EXPECT_DOUBLE_EQ(d::read_double(back.Points(), p * 3 + k),
                                 d::read_double(m.Points(), p * 3 + k));
        ASSERT_TRUE(back.HasPointData("f"));
        for (std::size_t p = 0; p < m.NumPoints(); ++p)
            EXPECT_DOUBLE_EQ(d::read_double(back.PointData("f"), p), static_cast<double>(p) * 0.5);
        ASSERT_TRUE(back.HasCellData("tag"));
        for (std::size_t c = 0; c < 27; ++c)
            EXPECT_EQ(d::read_int(back.CellData("tag", 0), c), static_cast<std::int64_t>(c));
        std::remove(path.c_str());
    }
}

#ifdef MESHIOPLUSPLUS_HAS_ZLIB
TEST(Vti, RoundTripsCompressed) {
    Mesh m = vti_lattice(4, 0.0, 1.0);
    NDArray pd = NDArray::Uninit(meshioplusplus::DType::Float64, {m.NumPoints()});
    for (std::size_t p = 0; p < m.NumPoints(); ++p)
        pd.As<double>()[p] = std::sin(static_cast<double>(p));
    m.AddPointData("f", std::move(pd));
    const std::string path = mt::temp_path(".vti");
    write_vti(path, m, /*binary=*/true, /*zlib=*/true);
    EXPECT_NE(vti_slurp(path).find("vtkZLibDataCompressor"), std::string::npos);
    const Mesh back = read_vti(path);
    for (std::size_t p = 0; p < m.NumPoints(); ++p)
        EXPECT_DOUBLE_EQ(d::read_double(back.PointData("f"), p),
                         d::read_double(m.PointData("f"), p));
    std::remove(path.c_str());
}
#endif

// The written bytes, not a read-back: a consistently wrong Origin/Spacing pair
// survives a round trip through this file's own reader unchanged.
TEST(Vti, TheWrittenAttributesAreTheGeometry) {
    const std::string path = mt::temp_path(".vti");
    write_vti(path, vti_lattice(3, -0.5, 0.25), /*binary=*/false, /*zlib=*/false);
    const std::string text = vti_slurp(path);
    std::remove(path.c_str());
    EXPECT_NE(text.find("type=\"ImageData\""), std::string::npos);
    EXPECT_NE(text.find("WholeExtent=\"0 3 0 3 0 3\""), std::string::npos);
    EXPECT_NE(text.find("Origin=\"-0.5 -0.5 -0.5\""), std::string::npos);
    EXPECT_NE(text.find("Spacing=\"0.25 0.25 0.25\""), std::string::npos);
    // No point array at all: the geometry IS the three attributes, and writing
    // points as well would make the file self-contradictory.
    EXPECT_EQ(text.find("<Points>"), std::string::npos);
}

// A hand-written file whose expected geometry is stated here rather than
// produced by our writer.
TEST(Vti, ReadsAForeignFileWithANonZeroOriginAndAShiftedExtent) {
    const std::string path = vti_write_temp(
        "<?xml version=\"1.0\"?>\n"
        "<VTKFile type=\"ImageData\" version=\"0.1\" byte_order=\"LittleEndian\">\n"
        "<ImageData WholeExtent=\"2 4 0 1 0 1\" Origin=\"1 2 3\" Spacing=\"0.5 2 4\">\n"
        "<Piece Extent=\"2 4 0 1 0 1\">\n"
        "<PointData><DataArray type=\"Int32\" Name=\"i\" format=\"ascii\">\n"
        "0 1 2 3 4 5 6 7 8 9 10 11\n</DataArray></PointData>\n"
        "</Piece></ImageData></VTKFile>\n");
    const Mesh m = read_vti(path);
    std::remove(path.c_str());

    // 2 x 1 x 1 cells, so 3 x 2 x 2 = 12 points.
    ASSERT_EQ(m.NumPoints(), 12u);
    ASSERT_EQ(m.NumCellBlocks(), 1u);
    EXPECT_EQ(m.Cells(0).NumCells(), 2u);
    // The extent starts at (2, 0, 0), so the lo corner is Origin + start*Spacing
    // = (1 + 2*0.5, 2, 3) = (2, 2, 3). Dropping that offset would translate the
    // whole grid by one cell in x -- silently, and plausibly.
    EXPECT_DOUBLE_EQ(d::read_double(m.Points(), 0), 2.0);
    EXPECT_DOUBLE_EQ(d::read_double(m.Points(), 1), 2.0);
    EXPECT_DOUBLE_EQ(d::read_double(m.Points(), 2), 3.0);
    // x fastest: point 1 is one spacing along x.
    EXPECT_DOUBLE_EQ(d::read_double(m.Points(), 3), 2.5);
    // point 3 is the first of the next y row.
    EXPECT_DOUBLE_EQ(d::read_double(m.Points(), 3 * 3 + 1), 4.0);
    ASSERT_TRUE(m.HasPointData("i"));
    EXPECT_EQ(d::read_int(m.PointData("i"), 11), 11);
}

TEST(Vti, MetadataAgreesWithTheRealRead) {
    const std::string path = mt::temp_path(".vti");
    Mesh m = vti_lattice(3, -0.5, 0.25);
    NDArray pd = NDArray(meshioplusplus::DType::Float64, {m.NumPoints()});
    m.AddPointData("f", std::move(pd));
    write_vti(path, m, /*binary=*/false, /*zlib=*/false);
    const meshioplusplus::MeshMetadata meta = read_vti_metadata(path);
    const Mesh back = read_vti(path);
    std::remove(path.c_str());

    EXPECT_EQ(meta.mNumPoints, back.NumPoints());
    ASSERT_EQ(meta.mCellBlocks.size(), 1u);
    EXPECT_EQ(meta.mCellBlocks[0].mType, "hexahedron");
    EXPECT_EQ(meta.mCellBlocks[0].mNumCells, back.Cells(0).NumCells());
    ASSERT_EQ(meta.mPointDataNames.size(), 1u);
    EXPECT_EQ(meta.mPointDataNames[0], "f");
    // Unlike every other native metadata path this one CAN report the bounding
    // box, because the extent is the bounding box.
    EXPECT_TRUE(meta.mHasBBox);
    EXPECT_DOUBLE_EQ(meta.mBBoxMin[0], -0.5);
    EXPECT_DOUBLE_EQ(meta.mBBoxMax[0], 0.25);
}

TEST(Vti, RefusesAMeshThatIsNotALattice) {
    const std::string path = mt::temp_path(".vti");
    // A tetra mesh has no extent at all.
    EXPECT_THROW(write_vti(path, mt::tet_mesh(), false, false), WriteError);
    // A PARTIAL lattice is the interesting refusal: it is made of the right
    // cells in the right places, and ImageData still cannot express its holes.
    meshioplusplus::VoxelOptions vo;
    vo.mResolution = std::array<std::int64_t, 3>{{4, 4, 4}};
    vo.mFill = meshioplusplus::VoxelFill::Surface;
    const Mesh partial = meshioplusplus::voxelize(mt::tri_mesh(), vo).mMesh;
    EXPECT_THROW(write_vti(path, partial, false, false), WriteError);
    std::remove(path.c_str());
}

TEST(Vti, DeclinesTheConstructsItDoesNotImplement) {
    const char* cases[] = {
        // appended data
        "<VTKFile type=\"ImageData\"><AppendedData encoding=\"raw\">_</AppendedData>"
        "<ImageData WholeExtent=\"0 1 0 1 0 1\"><Piece Extent=\"0 1 0 1 0 1\"/>"
        "</ImageData></VTKFile>",
        // two pieces
        "<VTKFile type=\"ImageData\"><ImageData WholeExtent=\"0 1 0 1 0 1\">"
        "<Piece Extent=\"0 1 0 1 0 1\"/><Piece Extent=\"0 1 0 1 0 1\"/>"
        "</ImageData></VTKFile>",
        // a piece that is not the whole extent
        "<VTKFile type=\"ImageData\"><ImageData WholeExtent=\"0 2 0 1 0 1\">"
        "<Piece Extent=\"0 1 0 1 0 1\"/></ImageData></VTKFile>",
        // a rotated lattice
        "<VTKFile type=\"ImageData\"><ImageData WholeExtent=\"0 1 0 1 0 1\" "
        "Direction=\"0 1 0 -1 0 0 0 0 1\"><Piece Extent=\"0 1 0 1 0 1\"/>"
        "</ImageData></VTKFile>",
        // an inverted extent
        "<VTKFile type=\"ImageData\"><ImageData WholeExtent=\"2 0 0 1 0 1\">"
        "<Piece Extent=\"2 0 0 1 0 1\"/></ImageData></VTKFile>",
        // the wrong dataset type
        "<VTKFile type=\"UnstructuredGrid\"><UnstructuredGrid/></VTKFile>",
    };
    for (const char* text : cases) {
        const std::string path = vti_write_temp(text);
        EXPECT_THROW(read_vti(path), ReadError) << text;
        std::remove(path.c_str());
    }
}

TEST(Vti, AnArrayOfTheWrongLengthIsAnErrorRatherThanATruncation) {
    const std::string path = vti_write_temp(
        "<VTKFile type=\"ImageData\"><ImageData WholeExtent=\"0 1 0 1 0 1\">"
        "<Piece Extent=\"0 1 0 1 0 1\">"
        "<PointData><DataArray type=\"Int32\" Name=\"i\" format=\"ascii\">1 2 3</DataArray>"
        "</PointData></Piece></ImageData></VTKFile>");
    EXPECT_THROW(read_vti(path), ReadError);
    std::remove(path.c_str());
}

// The identity the format exists for: written and read back, a grid IS the same
// grid -- geometry included, which no other format in meshio++ can promise for a
// mesh whose only description was `field_data`.
TEST(Vti, TheRecoveredHeaderEqualsTheWrittenOne) {
    const Mesh m = vti_lattice(5, 1.25, 0.125);
    d::LatticeSpec before;
    ASSERT_TRUE(d::lattice_from_mesh(m, before));

    const std::string path = mt::temp_path(".vti");
    write_vti(path, m, /*binary=*/true, /*zlib=*/false);
    const Mesh back = read_vti(path);
    std::remove(path.c_str());

    d::LatticeSpec after;
    ASSERT_TRUE(d::lattice_from_mesh(back, after));
    for (std::size_t k = 0; k < 3; ++k) {
        EXPECT_EQ(after.mDims[k], before.mDims[k]);
        EXPECT_DOUBLE_EQ(after.mOrigin[k], before.mOrigin[k]);
        EXPECT_DOUBLE_EQ(after.mSpacing[k], before.mSpacing[k]);
    }
}

TEST(LatticeFromMesh, RejectsWhatIsNotADenseLattice) {
    d::LatticeSpec spec;
    // A tetra mesh.
    EXPECT_FALSE(d::lattice_from_mesh(mt::tet_mesh(), spec));
    // A lattice whose points have been permuted: identical plane sets, different
    // mesh. Checking only the plane positions would accept this.
    Mesh m = vti_lattice(2, 0.0, 1.0);
    NDArray pts = NDArray::Uninit(meshioplusplus::DType::Float64, {m.NumPoints(), 3u});
    for (std::size_t p = 0; p < m.NumPoints(); ++p) {
        const std::size_t src = m.NumPoints() - 1 - p;
        for (std::size_t k = 0; k < 3; ++k)
            pts.As<double>()[p * 3 + k] = d::read_double(m.Points(), src * 3 + k);
    }
    Mesh permuted;
    permuted.AssignPoints(std::move(pts));
    permuted.AddCellBlock("hexahedron", m.Cells(0).Conn());
    EXPECT_FALSE(d::lattice_from_mesh(permuted, spec));
    // A graded grid: uniform plane counts, non-uniform spacing.
    Mesh graded;
    std::vector<double> xs = {0.0, 1.0, 3.0};
    NDArray gp = NDArray::Uninit(meshioplusplus::DType::Float64, {27u, 3u});
    std::size_t idx = 0;
    for (std::size_t k = 0; k < 3; ++k)
        for (std::size_t j = 0; j < 3; ++j)
            for (std::size_t i = 0; i < 3; ++i, ++idx) {
                gp.As<double>()[idx * 3 + 0] = xs[i];
                gp.As<double>()[idx * 3 + 1] = static_cast<double>(j);
                gp.As<double>()[idx * 3 + 2] = static_cast<double>(k);
            }
    graded.AssignPoints(std::move(gp));
    graded.AddCellBlock("hexahedron", vti_lattice(2, 0.0, 1.0).Cells(0).Conn());
    EXPECT_FALSE(d::lattice_from_mesh(graded, spec));
}
